ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java

```diff
@@ -244,7 +249,32 @@ private void removeFiles(String location, ValidWriteIdList writeIdList, Compacti
     obsoleteDirs.addAll(dir.getAbortedDirectories());
     List<Path> filesToDelete = new ArrayList<>(obsoleteDirs.size());
     StringBuilder extraDebugInfo = new StringBuilder("[");
+    boolean delayedCleanupEnabled = conf.getBoolVar(HiveConf.ConfVars.HIVE_COMPACTOR_DELAYED_CLEANUP_ENABLED);
+
     for (Path stat : obsoleteDirs) {
+      if (delayedCleanupEnabled) {
+        String filename = stat.toString();
+        if (filename.startsWith(AcidUtils.BASE_PREFIX)) {
+          long writeId = AcidUtils.ParsedBase.parseBase(stat).getWriteId();
+          if (ci.type == CompactionType.MINOR) {
+            LOG.info("Skipping base dir " + stat + " as this cleanup is for minor compaction"
+                + ", compaction id " + ci.id);
+            continue;
+          } else if (writeId > writeIdList.getHighWatermark()) {
+            LOG.info("Skipping base dir " + stat + " deletion as WriteId of this base dir is"
+                + " greater than highWaterMark for compaction id " + ci.id);
+            continue;
+          }
+        }
+        else if (filename.startsWith(AcidUtils.DELTA_PREFIX) || filename.startsWith(AcidUtils.DELETE_DELTA_PREFIX)) {
+          AcidUtils.ParsedDelta delta = AcidUtils.parsedDelta(stat, fs);
```

Comment:
There is a ParsedDeltaLight in AcidUtils, that is a cheaper way to parse out the maxWriteId. This parsedDelta method will issue a FS call to check if the delta is in raw format

Reply:
It would be helpful to extract this logic to a separate method.